### PR TITLE
[SYCL] Emit integration footer and header for device_global variables

### PR DESCRIPTION
**Pull request by Fznamznon**
The implementation is based on the design doc
https://github.com/intel/llvm/blob/sycl/sycl/doc/design/DeviceGlobal.md

### CODE REVIEW

**Comment by Fznamznon** [comment]
@erichkeane I'd appreciate review from you as well, since it seems that no one have never made any changes to integration footer except you.

**Comment by Fznamznon** [comment]
Review request for runtime seems like a bug.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [inline clang/lib/Sema/SemaDeclAttr.cpp:10419]: You can remove this codes. Add let SimpleHandler = 1; inside Attr.td file

File: clang/include/clang/Basic/Attr.td
Comment (smanna12) [inline clang/include/clang/Basic/Attr.td:1271]: Could you please update Documentation?

**Comment by bader** [comment]
> Review request for runtime seems like a bug.

# DPC++ runtime library
sycl/ @intel/llvm-reviewers-runtime

This rule matches any directory named "sycl" i.e. it's not necessary must be a root directory.
clang/test/CodeGenSYCL/Inputs/CL/sycl/detail/defines_elementary.hpp - matches this rule.

**Review by erichkeane — APPROVED** [review]
Seems alright to me!

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (Fznamznon) [reply clang/lib/Sema/SemaDeclAttr.cpp:10419]: I think more changes will come from @schittir for this attribute, so I'm not sure if it will stay as a simple handler. @schittir , WDYT?

Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:10419]: Sorry i did not know about more changes coming from @schittir. Based on your patch, i thought about simple handler.
You can ignore this one then

**Comment by Fznamznon** [comment]
> > Review request for runtime seems like a bug.
> 
> ```
> # DPC++ runtime library
> sycl/ @intel/llvm-reviewers-runtime
> ```
> 
> This rule matches any directory named "sycl" i.e. it's not necessary must be a root directory. clang/test/CodeGenSYCL/Inputs/CL/sycl/detail/defines_elementary.hpp - matches this rule.

Ok, thanks!

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (schittir) [reply clang/lib/Sema/SemaDeclAttr.cpp:10419]: Yes, please leave it as it is. I will update it.

Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:10419]: Thanks for checking.

File: clang/test/CodeGenSYCL/device_global_int_footer_header.cpp
Comment (erichkeane) [reply clang/test/CodeGenSYCL/device_global_int_footer_header.cpp:98]: Note to @Fznamznon : This command line in the check doesn't actually have a guid in it, so that might be why it doesn't show up, its probably worth giving that a shot.

Command line flag is -fsycl-unique-prefix=XXXX

The set of 4 underscores at the left side is I think associated with the unique prefix, the format is "<unique-prefix>____<rest>" (where those are 4 underscores).

File: clang/test/CodeGenSYCL/device_global_int_footer_header.cpp
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/device_global_int_footer_header.cpp:69]: D'oh! Will fix that.

File: clang/test/CodeGenSYCL/device_global_int_footer_header.cpp
Comment (erichkeane) [reply clang/test/CodeGenSYCL/device_global_int_footer_header.cpp:98]: FWIW, I would think adding it to this test is probably beneficial.

Comment (Fznamznon) [reply clang/test/CodeGenSYCL/device_global_int_footer_header.cpp]: Ok, done.

**Comment by Fznamznon** [comment]
Ping @intel/dpcpp-cfe-reviewers for review.

File: clang/include/clang/Sema/Sema.h
Comment (Fznamznon) [reply clang/include/clang/Sema/Sema.h:443]: I think something like `isDeviceGlobalsEmitted` will be a bit better since this function doesn't make emission. WDYT?

**Review by gmlueck — DISMISSED** [review]

**Review by premanandrao — COMMENTED** [review]
Some minor nits, but looks good to me.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:4694]: I'm not a big expert in English spellings, so, I kind of copy pasted it from the design doc - https://github.com/intel/llvm/blob/sycl/sycl/doc/design/DeviceGlobal.md .
@gmlueck , is that ok that we make this change?

Comment (gmlueck) [reply clang/lib/Sema/SemaSYCL.cpp:4694]: Yes, I like that name better.  Can you also update the sample code in the "DeviceGlobal.md" design document, so it stays in sync with the implementation?

**Review by elizabethandrews — COMMENTED** [review]
What happens when we capture a `device_global` variable with regard to openCL kernel generation. We decompose the type?

File: clang/test/CodeGenSYCL/device_global_int_footer_header.cpp
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/device_global_int_footer_header.cpp:9]: emitted constructs*

**Comment by Fznamznon** [comment]
> What happens when we capture a `device_global` variable with regard to openCL kernel generation. We decompose the type?

My understanding was that `device_global` variables won't be captured, because lambdas don't capture variables with static storage duration. `device_global` variables can be declared only at namespace scope or as a static member variable, but only if the member variable is publicly accessible from namespace scope. So `device_global` variables will have static storage duration.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:4694]: > Can you also update the sample code in the "DeviceGlobal.md" design document, so it stays in sync with the implementation?

Sure!

File: clang/lib/Sema/Sema.cpp
Comment (Fznamznon) [reply clang/lib/Sema/Sema.cpp]: Done, thanks!

File: clang/lib/Sema/Sema.cpp
Comment (Fznamznon) [reply clang/lib/Sema/Sema.cpp]: Done, thanks!

File: clang/include/clang/Sema/Sema.h
Comment (Fznamznon) [reply clang/include/clang/Sema/Sema.h]: Done, thanks!

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Done, thanks!

File: clang/test/CodeGenSYCL/device_global_int_footer_header.cpp
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/device_global_int_footer_header.cpp]: Done, thanks!

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:4978]: Renamed all the functions, thanks for pointing this out.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:5077]: Sure, I've tried to arrange the code this way in the latest commit. Please have a look.

File: clang/test/CodeGenSYCL/device_global_int_footer_header.cpp
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/device_global_int_footer_header.cpp]: Done, thanks!

File: clang/test/CodeGenSYCL/device_global_int_footer_header.cpp
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/device_global_int_footer_header.cpp]: Done, thanks!

**Comment by gmlueck** [comment]
> What happens when we capture a `device_global` variable with regard to openCL kernel generation. We decompose the type?

Are you asking what happens if a `device_global` is captured as a kernel argument?  This isn't possible because `device_global` variables must be declared either at namespace scope or as a static member variable in a class.  Variables of those types can't be captured by a lambda, so they aren't kernel arguments.  If the kernel is a named function object, only non-static member variables of the class become kernel arguments.

**Review by smanna12 — APPROVED** [review]

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (Fznamznon) [reply clang/lib/Sema/SemaDeclAttr.cpp:10419]: Yeah, after looking at @schittir's changes, I think a simple handler inside Attr.td file makes more sense.

**Review by premanandrao — DISMISSED** [review]
Changes look good to me.

**Review by elizabethandrews — DISMISSED** [review]
LGTM. Thank you!

**Comment by schittir** [comment]
Difference in naming of attributes between this patch and https://github.com/intel/llvm/pull/5597 
This one uses SYCLDeviceGlobal and SYCLGlobalVariableAllowed, whereas the latter uses SYCLDetailDeviceGlobal and SYCLDetailGlobalVariableAllowed

**Comment by Fznamznon** [comment]
Thanks @elizabethandrews, @smanna12, @premanandrao for reviews.
Although, I've made a small change adding simple handler to the attribute in Attr.td file and aligned mock declaration of `device_global` class with future changes coming from @schittir . Could you please take a look again and approve if that is ok?

**Comment by Fznamznon** [comment]
@alexbatashev , FYI. There are two jobs that are crashing on sycl-ls step.

**Review by smanna12 — APPROVED** [review]
LGTM

**Comment by bader** [comment]
> @alexbatashev , FYI. There are two jobs that are crashing on sycl-ls step.

Here too - https://github.com/intel/llvm/pull/5683. Can it be caused by recent GPU driver update?

**Review by premanandrao — DISMISSED** [review]
LGTM

File: clang/lib/Sema/SemaSYCL.cpp
Comment (schittir) [inline clang/lib/Sema/SemaSYCL.cpp:128]: Just FYI, I refactored isSyclDeviceGlobalType to be a member of Sema 
https://github.com/intel/llvm/pull/5597/commits/df337a7de8f9164bc63c7d82ed6e0f6120920aa5#diff-edac6256ac508912a16d0165b2f8cf37123dc2f40a147dca49a34c33f1db13dd

**Review by schittir — DISMISSED** [review]

**Review by elizabethandrews — DISMISSED** [review]

**Comment by Fznamznon** [comment]
I've reported failure with buildbot/Lit_With_Cuda . It is unrelated to PR changes.
Other failures is also unrelated (it is a crash on sycl-ls step).

**Comment by Fznamznon** [comment]
I can't remove the review request for @intel/llvm-reviewers-runtime . But there is no runtime changes.

**Review by bader — CHANGES_REQUESTED** [review]
Please, remove clang/test/CodeGenSYCL/Inputs/CL/sycl/detail/ directly and move all the files from this directory to clang/test/CodeGenSYCL/Inputs.
This will also resolve the problem with code owners.

**Comment by Fznamznon** [comment]
> Please, remove clang/test/CodeGenSYCL/Inputs/CL/sycl/detail/ directly and move all the files from this directory to clang/test/CodeGenSYCL/Inputs. This will also resolve the problem with code owners.

Without this directory I can't compile integration header and footer from clang tests. Do you suggest I remove this kind of testing?

**Comment by bader** [comment]
> > Please, remove clang/test/CodeGenSYCL/Inputs/CL/sycl/detail/ directly and move all the files from this directory to clang/test/CodeGenSYCL/Inputs. This will also resolve the problem with code owners.
> 
> Without this directory I can't compile integration header and footer from clang tests. Do you suggest to remove this kind of testing?

Are you saying that we hardcoded DPC++ runtime library headers directory path in clang sources? If so, it looks like a bug to me. I don't think clang must rely on header file locations and especially for "detail" functionality, which is not documented in the specification.

**Comment by Fznamznon** [comment]
> > > > Please, remove clang/test/CodeGenSYCL/Inputs/CL/sycl/detail/ directly and move all the files from this directory to clang/test/CodeGenSYCL/Inputs. This will also resolve the problem with code owners.
> > > 
> > > 
> > > Without this directory I can't compile integration header and footer from clang tests. Do you suggest to remove this kind of testing?
> > 
> > 
> > Are you saying that we hardcoded DPC++ runtime library headers directory path in clang sources? If so, it looks like a bug to me. I don't think clang must rely on header file locations and especially for "detail" functionality, which is not documented in the specification.
> 
> Yes, paths to some headers are hardcoded in emitted integration header and footer. It's been there for a long time, see -
> 
> https://github.com/intel/llvm/blob/0cff80eaa6e11179c276986b4b579147a9f2c414/clang/lib/Sema/SemaSYCL.cpp#L4626
> 
> .

I can file a bug request, but this is not a problem of this concrete change.

**Comment by bader** [comment]
> > > > Please, remove clang/test/CodeGenSYCL/Inputs/CL/sycl/detail/ directly and move all the files from this directory to clang/test/CodeGenSYCL/Inputs. This will also resolve the problem with code owners.
> > > 
> > > 
> > > Without this directory I can't compile integration header and footer from clang tests. Do you suggest to remove this kind of testing?
> > 
> > 
> > Are you saying that we hardcoded DPC++ runtime library headers directory path in clang sources? If so, it looks like a bug to me. I don't think clang must rely on header file locations and especially for "detail" functionality, which is not documented in the specification.
> 
> Yes, paths to some headers are hardcoded in emitted integration header and footer.

Okay.

> Do you suggest to remove this kind of testing?

Yes. I think checking the content of generated integration header/footer should be enough. Compilation of generated files is already covered by `check-sycl` test. Right?

**Comment by Fznamznon** [comment]
> Compilation of generated files is already covered by check-sycl test. Right?

I think this is not yet covered. I can't find declaration of `device_global` class in SYCL headers and it's usage in `sycl/test` directory, that means new content is not generated in check-sycl tests now and therefore compilation is not checked there. I assume that it should happen in the future when headers implement `device_global` though.

**Comment by bader** [comment]
> > Compilation of generated files is already covered by check-sycl test. Right?
> 
> I think this is not yet covered. I can't find declaration of `device_global` class in SYCL headers and it's usage in `sycl/test` directory, that means new content is not generated in check-sycl tests now and therefore compilation is not checked there. I assume that it should happen in the future when headers implement `device_global` though.

Anyway, the test suite for testing DPC++ headers is `check-sycl`, not `check-clang`.

**Comment by bader** [comment]
> > > > > Please, remove clang/test/CodeGenSYCL/Inputs/CL/sycl/detail/ directly and move all the files from this directory to clang/test/CodeGenSYCL/Inputs. This will also resolve the problem with code owners.
> > > > 
> > > > 
> > > > Without this directory I can't compile integration header and footer from clang tests. Do you suggest to remove this kind of testing?
> > > 
> > > 
> > > Are you saying that we hardcoded DPC++ runtime library headers directory path in clang sources? If so, it looks like a bug to me. I don't think clang must rely on header file locations and especially for "detail" functionality, which is not documented in the specification.
> > 
> > 
> > Yes, paths to some headers are hardcoded in emitted integration header and footer. It's been there for a long time, see -
> > https://github.com/intel/llvm/blob/0cff80eaa6e11179c276986b4b579147a9f2c414/clang/lib/Sema/SemaSYCL.cpp#L4626
> > 
> > .
> 
> I can file a bug, but this is not a problem of this concrete change.

I think we should do this. It looks like a design issue to me. We better avoid dependencies like this in the compiler. 

BTW, it looks like it should be easy to do. https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/defines_elementary.hpp defines some number of macro. I think we should just expand all of them in emitted integration header and remove this include.


`O << "#include <CL/sycl/detail/defines_elementary.hpp>\n"; `

**Comment by Fznamznon** [comment]
> > > > > > Please, remove clang/test/CodeGenSYCL/Inputs/CL/sycl/detail/ directly and move all the files from this directory to clang/test/CodeGenSYCL/Inputs. This will also resolve the problem with code owners.
> > > > > 
> > > > > 
> > > > > Without this directory I can't compile integration header and footer from clang tests. Do you suggest to remove this kind of testing?
> > > > 
> > > > 
> > > > Are you saying that we hardcoded DPC++ runtime library headers directory path in clang sources? If so, it looks like a bug to me. I don't think clang must rely on header file locations and especially for "detail" functionality, which is not documented in the specification.
> > > 
> > > 
> > > Yes, paths to some headers are hardcoded in emitted integration header and footer. It's been there for a long time, see -
> > > https://github.com/intel/llvm/blob/0cff80eaa6e11179c276986b4b579147a9f2c414/clang/lib/Sema/SemaSYCL.cpp#L4626
> > > 
> > > .
> > 
> > 
> > I can file a bug, but this is not a problem of this concrete change.
> 
> I think we should do this. It looks like a design issue to me. We better avoid dependencies like this in the compiler.
> 
> BTW, it looks like it should be easy to do. https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/defines_elementary.hpp defines some number of macro. I think we should just expand all of them in emitted integration header and remove this include.
> 
> `O << "#include <CL/sycl/detail/defines_elementary.hpp>\n"; `

It is just one of them. I think it is not that simple for others. I've submitted an internal tracker.

**Review by bader — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Comment by Fznamznon** [comment]
I don't think I broke `L0 GEN9 LLVM Test Suite (pull_request) ` this time either. The latest commit is different from previous one (which passed this check) only by changes in clang LIT tests. Some part of the tests are failing on `sycl-ls` command.

**Comment by premanandrao** [comment]
@Fznamznon, I have a concern with the removal of footer and header compilation tests on host.   I think it leaves a hole in our testing, for example, if there is a typo or bad code generated in the header and footer, we won't notice it.  If there are tests elsewhere that would catch this, that would be sufficient, but without one we have a gap in coverage.

This happened with a recent PR, accidental missing quotes slipped through on first review.

**Comment by Fznamznon** [comment]
> I have a concern with the removal of footer and header compilation tests on host. I think it leaves a hole in our testing, for example, if there is a typo or bad code generated in the header and footer, we won't notice it.

Yeah, I had the same thoughts when did compilation testing. I didn't believe to my eyes and my c++ knowledge enough, so I added compiler's check. In fact, I even caught missing conversion to `void *` in integration footer emission. But if there were tests in `check-sycl` with `device_global`, they would have caught it.

> If there are tests elsewhere that would catch this, that would be sufficient, but without one we have a gap in coverage.

As @bader pointed out, `check-sycl` suite is supposed to perform host compilation and catch this. I'm ok with that.

**Comment by bader** [comment]
> I don't think I broke `L0 GEN9 LLVM Test Suite (pull_request) ` this time either. The latest commit is different from previous one (which passed this check) only by changes in clang LIT tests. Some part of the tests are failing on `sycl-ls` command.

@Fznamznon, could you create a tracker to investigate this problem? I suggest re-running the tests to check if this issue is sporadic.

**Comment by premanandrao** [comment]
> As @bader pointed out, `check-sycl` suite is supposed to perform host compilation and catch this. I'm ok with that.

Sounds good.

**Comment by Fznamznon** [comment]
> @Fznamznon, could you create a tracker to investigate this problem?

@bader , @alexbatashev , this is an issue with GitHub actions. Should I create a GitHub issue or an internal tracker?

**Comment by alexbatashev** [comment]
> > @Fznamznon, could you create a tracker to investigate this problem?
> 
> @bader , @alexbatashev , this is an issue with GitHub actions. Should I create a GitHub issue or an internal tracker? BTW, #5592 also experiences the same problem.

I don't see it as a GitHub Actions problem. I see that a few tests have failed for unrelated changes. Please, create internal tracker for those issues and disable those tests.

**Comment by Fznamznon** [comment]
> I don't see it as a GitHub Actions problem. 
Sorry, my bad.
> Please, create internal tracker for those issues and disable those tests.
Done. Created https://github.com/intel/llvm-test-suite/pull/886 .
It seems all recent PRs are affected by this fail.

**Pull request merged**
